docs/RFCS/sql_column_groups.md

```diff
@@ -0,0 +1,122 @@
+- Feature Name: sql_column_groups
+- Status: draft
+- Start Date: 2015-12-14
+- RFC PR:
+- Cockroach Issue:
+
+# Summary
+
+Store multiple columns for a row in a single value to amortize the
+overhead of keys and the fixed costs associated with values. The
+mapping of a column to a column group will be determined when a column
+is added to the table.
+
+# Motivation
+
+The SQL to KV mapping currently utilizes a key/value pair per
+non-primary key column. The key is structured as:
+
+```
+/<tableID>/<indexID>/<primaryKeyColumns...>/<columnID>
+```
+
+Keys also have an additional suffix imposed by the MVCC
+implementation: ~10 bytes of timestamp.
+
+The value is a proto (note, this is what is stored on disk, we also
+include a timestamp when transmitting over the network):
+
+```
+message Value {
+  optional bytes raw_bytes;
+  optional fixed32 checksum;
+  optional ValueType tag;
+}
+```
+
+Consider the following table:
+
+```
+CREATE TABLE kv (
+  k INT PRIMARY KEY,
+  v INT
+)
+```
+
+A row for this table logically has 16 bytes of data (`INTs` are 64-bit
+values). The current SQL to KV mapping will create 2 key/value pairs:
+a sentinel key for the row and the key/value for column `v`. These two
+keys imply an overhead of ~28 bytes (timestamp + checksum) when stored
+on disk and more than that when transmitted over the network (disk
+storage takes advantage of prefix compression of the keys). We can cut
+this overhead in half by storing only a single key/value for the
+row. The savings are even more significant as more columns are added
+to the table.
+
+Note that there is also per key overhead at the transaction
+level. Every key written within a transaction has a "write intent"
+associated with it and these intents need to be resolved when the
+transaction is committed. Reducing the number of keys per row reduces
+this overhead.
+
+# Detailed design
+
+The structure for table keys will be changed to:
+
+```
+/<tableID>/<indexID>/<primaryKeyColumns...>/<groupID>
+```
+
+The value proto will remain almost the same, but the `tag` field will
+be removed and instead of encoding the column value directly into the
+`raw_bytes` field we'll encoding a series of `<columnID>/<columnVal>`
```

Comment:
Yes, it is a micro-optimization, but a significant one. We're already partially doing our own encoding of the value in `raw_bytes` rather than having some sort of proto union type. I've also thought it might be better to get rid of the Value structure altogether and use a fixed-length checksum followed by the raw string. Another advantage is that we could verify the checksum without decoding and re-encoding the proto as is currently done at the transition from the MVCC layer to the Store layer.